File: BoneCentral/Brain/CppInterface/pid_controller.js
module.exports = (function() {

    function PidController() {
        this._previousError = 0;
        this._min = -100;
        this._max = 100;
        this._kp = 0.1;
        this._ki = 0.5;
        this._kd = 0.01;
    }

    PidController.prototype.withBounds = function(min, max) {
        this._min = min;
        this._max = max;
    };

    PidController.prototype.withTimeDelta = function (timeDelta) {
        this._timeDelta = timeDelta;
    };
    
    PidController.prototype.calculateAdjustment = function(setPoint, currentMeasurement) {
        var currentError = setPoint - currentMeasurement;
        var errorDifference = currentError-this._previousError;
        this._previousError = currentError;
        var proportion  = this._kp * currentError;
        var integral  = this._ki * currentError * this._timeDelta;
        var derivative  = this._kd * (errorDifference)/this._timeDelta;
        var adjustment = proportion + integral + derivative;
        if(adjustment > this._max)
            return this._max;
        if(adjustment < this._min)
            return this._min;
        return adjustment;
    };
    
    PidController.prototype.calibrate = function(p, i, d) {
        this._kp = p;
        this._ki = i;
        this._kd = d;
    };

    PidController.prototype.reset = function() {
        this._previousError = 0;
    };
    
    return PidController;

})();

File: BoneCentral/Peripherals/Peripherals.Core/src/PidController.cpp
//
// Created by nfcop on 7/15/2017.
//

#include "PidController.h"

PidController& PidController::withConstants(Configuration configuration) {
    p_ = configuration.P;
    i_ = configuration.I;
    d_ = configuration.D;
    return *this;
}

PidController& PidController::withBounds(float minimum, float maximum) {
    min_ = minimum; max_ = maximum;
    return *this;
}

PidController& PidController::withTimeDelta(unsigned int timeDelta) {
    timeDelta_ = timeDelta;
    return *this;
}

PidController& PidController::updateSetPoint(float setPoint) {
    std::lock_guard<std::mutex> lock(setPointMutex_);
    setPoint_ = setPoint;
    return *this;
}

float PidController::calculateAdjustmentFor(float setPoint, float processValue) {
    auto currentError = setPoint - processValue;
    auto errorDifference = currentError - previousError_;
    previousError_ = currentError;
    integral_  = integral_ + currentError * timeDelta_;
    auto derivative  = errorDifference / timeDelta_;
    auto adjustment = p_ * currentError + i_ * integral_ + d_ * derivative;
    if(adjustment > max_) return max_;
    if(adjustment < min_) return min_;
    return adjustment;
}

PidController& PidController::start() {
    if (thread_ != nullptr) return *this;
    //setNeutral();
    stop_ = false;
    thread_ = new std::thread([&](){
        run();
    });
	return *this;
}

PidController& PidController::stop() {
    stop_ = true;
    thread_->join();
    delete thread_;
    thread_ = nullptr;
	return *this;
}

void PidController::run() {
    while (!stop_) {
		float curSetPoint = 0.0f;
        float processValue = getProcessValue();
        {
            std::lock_guard<std::mutex> lock(setPointMutex_);
            curSetPoint = setPoint_;
        }
        float power = calculateAdjustmentFor(curSetPoint,processValue);
        setControlValue(power);
        auto deltaInMicroseconds = timeDelta_ * 1000;
        usleep( deltaInMicroseconds );
    }
    setControlValue(0);
}
